src/grouping_map.rs

```diff
@@ -609,6 +565,484 @@ where
     where
         V: Mul<V, Output = V>,
     {
-        self.reduce(|acc, _, val| acc * val)
+        self.product_in(HashMap::new())
+    }
+}
+
+impl<I, K, V> GroupingMap<I>
+where
+    I: Iterator<Item = (K, V)>,
+{
+    /// Apply [`aggregate`](Self::aggregate) with a provided empty map
+    /// (`BTreeMap` or `HashMap` with any hasher).
+    ///
+    /// ```
+    /// use std::collections::BTreeMap;
+    /// use itertools::Itertools;
+    ///
+    /// let data = vec![2, 8, 5, 7, 9, 0, 4, 10];
+    /// let lookup = data.into_iter()
+    ///     .into_grouping_map_by(|&n| n % 4)
+    ///     .aggregate_in(|acc, _key, val| {
+    ///         if val == 0 || val == 10 {
+    ///             None
+    ///         } else {
+    ///             Some(acc.unwrap_or(0) + val)
+    ///         }
+    ///     }, BTreeMap::new());
+    ///
+    /// assert_eq!(lookup[&0], 4);        // 0 resets the accumulator so only 4 is summed
+    /// assert_eq!(lookup[&1], 5 + 9);
+    /// assert_eq!(lookup.get(&2), None); // 10 resets the accumulator and nothing is summed afterward
+    /// assert_eq!(lookup[&3], 7);
+    /// assert_eq!(lookup.len(), 3);      // The final keys are only 0, 1 and 2
+    /// ```
+    pub fn aggregate_in<FO, R, M>(self, mut operation: FO, mut map: M) -> M
+    where
+        FO: FnMut(Option<R>, &K, V) -> Option<R>,
+        M: Map<Key = K, Value = R>,
+    {
+        debug_assert!(map.is_empty());
+
+        self.iter.for_each(|(key, val)| {
+            let acc = map.remove(&key);
+            if let Some(op_res) = operation(acc, &key, val) {
+                map.insert(key, op_res);
+            }
```

Comment:
Since these two are the only uses of `remove` and `insert` you could replace them in the trait with the combined operation, which should also be implementable in a more efficient way for some types (e.g. `Vec<(K, V)>` but also `hashbrown`'s `HashMap` if that will ever be added)

Reply:
Good point. But I might prefer to optimize later as `Map` is meant to be used for way more than `GroupingMap` and I would prefer a larger point of view first. It won't be long (honestly `GroupingMap` is the biggest chunk for it).

Currently, `Map` is implemented for `std::collections::HashMap` so `hashbrown::HashMap` (which is not an alias of it) is not supported. What would be nice is `HashMap` to be moved to `alloc` but we are not there yet.